staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go

```diff
@@ -174,14 +175,43 @@ func (a customResourceStrategy) Validate(ctx context.Context, obj runtime.Object
 
 		// validate x-kubernetes-validations rules
 		if celValidator, ok := a.celValidators[v]; ok {
-			err, _ := celValidator.Validate(ctx, nil, a.structuralSchemas[v], u.Object, cel.RuntimeCELCostBudget)
-			errs = append(errs, err...)
+			// if contains OpenAPIv3 type/maxLength/maxItems/required/wrong type field validation failures, skip cel validation
+			if hasBlockingErr(errs) {
+				errs = append(errs, field.Invalid(nil, nil, "Blocking error occurred above, skip CEL validation"))
+			} else {
+				err, _ := celValidator.Validate(ctx, nil, a.structuralSchemas[v], u.Object, cel.RuntimeCELCostBudget)
+				errs = append(errs, err...)
+			}
 		}
 	}
 
 	return errs
 }
 
+func hasBlockingErr(errs field.ErrorList) bool {
+	// OpenAPIv3 type/maxLength/maxItems/required/wrong type field validation failures are viewed as blocking err for CEL validation
+	for _, err := range errs {
+		if err.Type == field.ErrorTypeInvalid && strings.Contains(err.Error(), "is an invalid type name") ||
+			err.Type == field.ErrorTypeInvalid && strings.Contains(err.Error(), "must be of type") ||
+			err.Type == field.ErrorTypeInvalid && strings.Contains(err.Error(), "must not be empty for specified object fields") ||
+			err.Type == field.ErrorTypeRequired ||
+			err.Type == field.ErrorTypeInvalid && containSubString(err.Error(), "should be at most", "chars long") ||
+			err.Type == field.ErrorTypeInvalid && containSubString(err.Error(), "should have at most", "items") {
```

Comment:
Worth adding a sub-type (e.g. `err.SubType` == field.ErrorSubTypeMaxItemsExceeded`) to all field.Error so we can check it and avoid the contains checks? It's going to be tedious to catalog them, but maybe we don't need to catalog them all.

Reply:
FWIW, I care more about maintainability than performance here right now (If there are no errors this has no performance impact)

Reply:
The maxLength/maxItems/maxProperties validation come from schema validation from kube-openapi. We might need to add there as well if wanna distinguish specific error.

Reply:
I think I can live with doing the contains checks here for alpha since it only incurs a cost if there are errors. Maybe open a issue so we fix this before promoting to beta? WDYT @liggitt?

Reply:
After revisit, we have the required error code in [kube-openapi](https://github.com/kubernetes/kube-openapi/blob/6a7b7046eec8fc952b8bbfbee4a2d581dbf85bdb/pkg/validation/errors/schema.go#L71). I will update later

Reply:
big +1 for using deterministic codes if they're plumbed to where we need the info (e.g. InvalidTypeCode, TooLongFailCode, MaxItemsFailCode, TooManyPropertiesCode)

if we don't have that info, I'd focus on the following:
* performance (avoid repeatedly calling `String()` on field name and error... just do it once each)
* correctness (is prefix on the field name correct or will it pick up false positives like a required field `.foo.bar.baz` keeping us from running a rule under `.foo.bar.bazz`?
* test coverage (put each type of error we're short-circuiting on in a distinct if block and make sure unit tests get 100% coverage of the `return true` cases)

Reply:
I have updated to use deterministic codes for the error